File: creational/dart/singleton_structured/database_connection.dart
class DatabaseConnection {
  static DatabaseConnection? _instance;
  
  DatabaseConnection._internal() {
    print('DatabaseConnection: Initializing connection...');
    _isConnected = true;
  }
  
  factory DatabaseConnection() {
    _instance ??= DatabaseConnection._internal();
    return _instance!;
  }
  
  bool _isConnected = false;
  int _queryCount = 0;
  
  void executeQuery(String query) {
    if (_isConnected) {
      _queryCount++;
      print('Executing query #$_queryCount: $query');
    } else {
      print('Error: Not connected to database');
    }
  }
  
  int get queryCount => _queryCount;
  
  bool get connectionStatus => _isConnected;
}

File: creational/typescript/singleton_structured/database_connection.ts
export class DatabaseConnection {
  private static instance: DatabaseConnection | null = null;
  
  private isConnected: boolean = false;
  private queryCount: number = 0;
  
  private constructor() {
    console.log('DatabaseConnection: Initializing connection...');
    this.isConnected = true;
  }
  
  public static getInstance(): DatabaseConnection {
    if (!DatabaseConnection.instance) {
      DatabaseConnection.instance = new DatabaseConnection();
    }
    return DatabaseConnection.instance;
  }
  
  public executeQuery(query: string): void {
    if (this.isConnected) {
      this.queryCount++;
      console.log(`Executing query #${this.queryCount}: ${query}`);
    } else {
      console.log('Error: Not connected to database');
    }
  }
  
  public getQueryCount(): number { return this.queryCount; }
  public getConnectionStatus(): boolean { return this.isConnected; }
}
